core/src/main/java/com/linecorp/armeria/internal/common/HttpHeadersUtil.java

```diff
@@ -88,9 +88,18 @@ public static RequestHeaders mergeRequestHeaders(RequestHeaders headers,
                                                      HttpHeaders defaultHeaders,
                                                      HttpHeaders additionalHeaders,
                                                      HttpHeaders internalHeaders) {
-        if (defaultHeaders.isEmpty() && additionalHeaders.isEmpty() && internalHeaders.isEmpty() &&
-            headers.contains(HttpHeaderNames.USER_AGENT)) {
-            return headers;
+        if (defaultHeaders.isEmpty() && additionalHeaders.isEmpty()) {
```

Comment:
Please add the `internalHeaders.isEmpty()` check back; we can still skip the loop completely by checking it.